### DISCUSSION TITLE
CH32V203 support

### DISCUSSION THREAD

**Discussion by drom** [hathach/tinyusb#2321] [Ideas]
I see that TinyUSB supports CH32V307. Anybody tried porting it to CH32V203?

**Comment by Krasutski** [comment]
All USB peripherals are the same for all CH32F2x, CH32V2x, and CH32V3x family.  The current dcd_ch32_usbhs.c Interface should work for all mentioned MCUs
![image](https://github.com/hathach/tinyusb/assets/52926632/eeafed5c-5142-4053-8fdd-fa4ba494f019)

**Comment by greenscreenflicker** [comment]
@Krasutski @drom I tryied your recommendation. But I run into a problem:
../tinyusb/portable/wch/dcd_ch32_usbhs.c: In function 'dcd_init':
../tinyusb/portable/wch/dcd_ch32_usbhs.c:69:5: error: 'USBHSD' undeclared (first use in this function)

Searching for USBHSD doesn't yield any results. Recommendations?

**Reply by Krasutski** [reply]
You need to pull MCU libs, if you use CH32V203 you need to add support for new MCU, please check my [similar PR](https://github.com/hathach/tinyusb/pull/2278/commits/ce627f431831e1fbaa3bb7bad35a489b8006b749)

**Reply by greenscreenflicker** [reply]
@Krasutski I'm not sure if i understood you correctly. Thanks for your support so far.
First, I have the MCU file headers included. But for instance, I don't have a USBHS_IRQn.
Eg. NVIC_EnableIRQ(USBHS_IRQn); does not exist as defined.
Instead there exists a USBHD_IRQn. E.g. NVIC_EnableIRQ(USBHD_IRQn); works.

Or let's take USBHSD. It's just undeclared. So the registers must be named differently, right?
Or did i just not include a file?
If the files are named differently? Is there an easy approach for porting?

I'm so stuck, please advise.

**Reply by greenscreenflicker** [reply]
@Krasutski I tried it with your commit the F20X for a F20X MCU, and it compiles. If it works, i cannot test due to a lag of test board.

**Comment by greenscreenflicker** [comment]
It seems that there's some definitions missing in CH32V20x.
I added them manually to dcd_ch32_usbhs.c

> #define USBHS_BASE            (AHBPERIPH_BASE + 0x3400)
/* USBHS Registers */
typedef struct
{
  __IO uint8_t  CONTROL;
  __IO uint8_t  HOST_CTRL;
  __IO uint8_t  INT_EN;
  __IO uint8_t  DEV_AD;
  __IO uint16_t FRAME_NO;
  __IO uint8_t  SUSPEND;
  __IO uint8_t  RESERVED0;
  __IO uint8_t  SPEED_TYPE;
  __IO uint8_t  MIS_ST;
  __IO uint8_t  INT_FG;
  __IO uint8_t  INT_ST;
  __IO uint16_t RX_LEN;
  __IO uint16_t RESERVED1;
  __IO uint32_t ENDP_CONFIG;
  __IO uint32_t ENDP_TYPE;
  __IO uint32_t BUF_MODE;
  __IO uint32_t UEP0_DMA;
  __IO uint32_t UEP1_RX_DMA;
  __IO uint32_t UEP2_RX_DMA;
  __IO uint32_t UEP3_RX_DMA;
  __IO uint32_t UEP4_RX_DMA;
  __IO uint32_t UEP5_RX_DMA;
  __IO uint32_t UEP6_RX_DMA;
  __IO uint32_t UEP7_RX_DMA;
  __IO uint32_t UEP8_RX_DMA;
  __IO uint32_t UEP9_RX_DMA;
  __IO uint32_t UEP10_RX_DMA;
  __IO uint32_t UEP11_RX_DMA;
  __IO uint32_t UEP12_RX_DMA;
  __IO uint32_t UEP13_RX_DMA;
  __IO uint32_t UEP14_RX_DMA;
  __IO uint32_t UEP15_RX_DMA;
  __IO uint32_t UEP1_TX_DMA;
  __IO uint32_t UEP2_TX_DMA;
  __IO uint32_t UEP3_TX_DMA;
  __IO uint32_t UEP4_TX_DMA;
  __IO uint32_t UEP5_TX_DMA;
  __IO uint32_t UEP6_TX_DMA;
  __IO uint32_t UEP7_TX_DMA;
  __IO uint32_t UEP8_TX_DMA;
  __IO uint32_t UEP9_TX_DMA;
  __IO uint32_t UEP10_TX_DMA;
  __IO uint32_t UEP11_TX_DMA;
  __IO uint32_t UEP12_TX_DMA;
  __IO uint32_t UEP13_TX_DMA;
  __IO uint32_t UEP14_TX_DMA;
  __IO uint32_t UEP15_TX_DMA;
  __IO uint16_t UEP0_MAX_LEN;
  __IO uint16_t RESERVED2;
  __IO uint16_t UEP1_MAX_LEN;
  __IO uint16_t RESERVED3;
  __IO uint16_t UEP2_MAX_LEN;
  __IO uint16_t RESERVED4;
  __IO uint16_t UEP3_MAX_LEN;
  __IO uint16_t RESERVED5;
  __IO uint16_t UEP4_MAX_LEN;
  __IO uint16_t RESERVED6;
  __IO uint16_t UEP5_MAX_LEN;
  __IO uint16_t RESERVED7;
  __IO uint16_t UEP6_MAX_LEN;
  __IO uint16_t RESERVED8;
  __IO uint16_t UEP7_MAX_LEN;
  __IO uint16_t RESERVED9;
  __IO uint16_t UEP8_MAX_LEN;
  __IO uint16_t RESERVED10;
  __IO uint16_t UEP9_MAX_LEN;
  __IO uint16_t RESERVED11;
  __IO uint16_t UEP10_MAX_LEN;
  __IO uint16_t RESERVED12;
  __IO uint16_t UEP11_MAX_LEN;
  __IO uint16_t RESERVED13;
  __IO uint16_t UEP12_MAX_LEN;
  __IO uint16_t RESERVED14;
  __IO uint16_t UEP13_MAX_LEN;
  __IO uint16_t RESERVED15;
  __IO uint16_t UEP14_MAX_LEN;
  __IO uint16_t RESERVED16;
  __IO uint16_t UEP15_MAX_LEN;
  __IO uint16_t RESERVED17;
  __IO uint16_t UEP0_TX_LEN;
  __IO uint8_t  UEP0_TX_CTRL;
  __IO uint8_t  UEP0_RX_CTRL;
  __IO uint16_t UEP1_TX_LEN;
  __IO uint8_t  UEP1_TX_CTRL;
  __IO uint8_t  UEP1_RX_CTRL;
  __IO uint16_t UEP2_TX_LEN;
  __IO uint8_t  UEP2_TX_CTRL;
  __IO uint8_t  UEP2_RX_CTRL;
  __IO uint16_t UEP3_TX_LEN;
  __IO uint8_t  UEP3_TX_CTRL;
  __IO uint8_t  UEP3_RX_CTRL;
  __IO uint16_t UEP4_TX_LEN;
  __IO uint8_t  UEP4_TX_CTRL;
  __IO uint8_t  UEP4_RX_CTRL;
  __IO uint16_t UEP5_TX_LEN;
  __IO uint8_t  UEP5_TX_CTRL;
  __IO uint8_t  UEP5_RX_CTRL;
  __IO uint16_t UEP6_TX_LEN;
  __IO uint8_t  UEP6_TX_CTRL;
  __IO uint8_t  UEP6_RX_CTRL;
  __IO uint16_t UEP7_TX_LEN;
  __IO uint8_t  UEP7_TX_CTRL;
  __IO uint8_t  UEP7_RX_CTRL;
  __IO uint16_t UEP8_TX_LEN;
  __IO uint8_t  UEP8_TX_CTRL;
  __IO uint8_t  UEP8_RX_CTRL;
  __IO uint16_t UEP9_TX_LEN;
  __IO uint8_t  UEP9_TX_CTRL;
  __IO uint8_t  UEP9_RX_CTRL;
  __IO uint16_t UEP10_TX_LEN;
  __IO uint8_t  UEP10_TX_CTRL;
  __IO uint8_t  UEP10_RX_CTRL;
  __IO uint16_t UEP11_TX_LEN;
  __IO uint8_t  UEP11_TX_CTRL;
  __IO uint8_t  UEP11_RX_CTRL;
  __IO uint16_t UEP12_TX_LEN;
  __IO uint8_t  UEP12_TX_CTRL;
  __IO uint8_t  UEP12_RX_CTRL;
  __IO uint16_t UEP13_TX_LEN;
  __IO uint8_t  UEP13_TX_CTRL;
  __IO uint8_t  UEP13_RX_CTRL;
  __IO uint16_t UEP14_TX_LEN;
  __IO uint8_t  UEP14_TX_CTRL;
  __IO uint8_t  UEP14_RX_CTRL;
  __IO uint16_t UEP15_TX_LEN;
  __IO uint8_t  UEP15_TX_CTRL;
  __IO uint8_t  UEP15_RX_CTRL;
} USBHSD_TypeDef;
#define USBHSD              ((USBHSD_TypeDef *) USBHS_BASE)

**Reply by Krasutski** [reply]
The current version doesn't support for CH32V20x but it is easy to integrate, similar as I did with ch32f2x.

**Reply by greenscreenflicker** [reply]
@Krasutski Would you be willing to do a pull request paid?

**Reply by Krasutski** [reply]
No need to pay but I can do it in the next days maybe on the weekends
also, I don't have a board to double-check all so be ready for new issues :)

**Reply by greenscreenflicker** [reply]
@Krasutski Thank you for your offer. I'm willing to send you my evalboard for free by express. Or I test it myself, as you please. 
![grafik](https://github.com/hathach/tinyusb/assets/17407241/7c31b785-5c6c-4014-8a45-b887962b00da)

If you wish to have the eval-board, we need to figure out a way to send me your address privately.

I personally don't feel to be a good-enough programmer for such pull requests.

**Reply by greenscreenflicker** [reply]
Hi, would be greatly happy about a pull request / support from your side. It turns out, that there are a lot of compability issues, as functions are missing in the wch definitions...
Let me know, as soon as I could try there something..

**Reply by Krasutski** [reply]
Just checked CH32V2x and as I see it doesn't support High-Speed USB, my PR was about HS driver, You need to use full-speed driver. The full-speed driver is under review now, you can pull this branch and test it: https://github.com/hathach/tinyusb/pull/2362

**Reply by greenscreenflicker** [reply]
Thank you for pointing me to that. Just tried to implement it. I have some issues there.

**Comment by hathach** [comment]
both usbfs and fsdev is supported by latest master.

**Comment by 203Null** [comment]
I see some work in the V203 host mode support in the repo. Are there still some works needed to be done? I could contribute to getting it to work.
